tests/datasets/test_utils_single_cell.py

```diff
@@ -0,0 +1,287 @@
+    adata_obj.var_names = var_names
+    adata_obj.obs_names = obs_names
+    adata_obj.obs["condition"] = pd.Categorical(conditions)
+
+    # Gene-specific expressions for filters
+    adata_obj.X[10:16, 4] = 0.0  # gene_4 zero in controls (NC)
+    adata_obj.X[:, 2] = 0.0
+    adata_obj.X[0, 2] = 1.0  # one A cell expresses gene_2
+    adata_obj.X[10, 2] = 1.0  # one control cell expresses gene_2
+
+    # Controls mapping for each condition
+    control_map = {
+        "A": ["cell_010", "cell_011", "cell_012"],
+        "A_small": ["cell_011", "cell_012", "cell_013"],
+        "B": ["cell_012", "cell_013", "cell_014"],
+    }
+    return adata_obj, control_map
+
+
+class TestCreateAdataForCondition:
+    @pytest.mark.parametrize(
+        "condition,signal_gene_idx,signal_value",
+        [("A", 0, 5.0), ("B", 1, 3.0)],
+    )
+    def test_shapes_and_labels(
+        self, make_adata, condition, signal_gene_idx, signal_value
+    ):
+        adata_obj, control_map = make_adata
+
+        # condition = "A"
+        condition_key = "condition"
+        control_name = "NC"
+
+        rows_cond = np.where(adata_obj.obs[condition_key] == condition)[0]
+        rows_ctrl = adata_obj.obs.index.get_indexer_for(control_map[condition])
+
+        merged, num_condition = create_adata_for_condition(
+            adata=adata_obj,
+            condition=condition,
+            condition_key=condition_key,
+            control_name=control_name,
+            rows_cond=rows_cond,
+            rows_ctrl=rows_ctrl,
+        )
+
+        assert isinstance(merged, ad.AnnData)
+        assert num_condition == len(rows_cond)
+        assert merged.n_obs == len(rows_cond) + len(rows_ctrl)
+        assert merged.n_vars == adata_obj.n_vars
+
+        labels = merged.obs[condition_key].tolist()
+        assert labels == [condition] * len(rows_cond) + [
+            f"{control_name}_{condition}"
+        ] * len(rows_ctrl)
+
+        assert np.allclose(merged.X[: len(rows_cond), signal_gene_idx], signal_value)
+        assert np.allclose(merged.X[len(rows_cond) :, signal_gene_idx], 1.0)
+
+    def test_logs_warning_on_length_mismatch(self, make_adata, caplog):
+        adata_obj, control_map = make_adata
+
+        condition = "A_small"  # 2 condition cells vs 3 controls
+        condition_key = "condition"
+        control_name = "NC"
+        rows_cond = np.where(adata_obj.obs[condition_key] == condition)[0]
+        rows_ctrl = adata_obj.obs.index.get_indexer_for(control_map[condition])
+
+        with caplog.at_level("WARNING"):
+            _ = create_adata_for_condition(
+                adata=adata_obj,
+                condition=condition,
+                condition_key=condition_key,
+                control_name=control_name,
+                rows_cond=rows_cond,
+                rows_ctrl=rows_ctrl,
+            )
+
+        assert any("different lengths" in rec.message for rec in caplog.records)
+
+
+class TestRunMulticonditionDGEAnalysis:
```

Comment:
Reviewer note: This was incorrectly placed in `tests/datasets/test_utils.py` and was moved to the correct location of `tests/datasets/test_utils_single_cell.py`